### PR TITLE
feat: add unit tests for authentication middelware 

### PR DESCRIPTION
**Pull request by MayankBansal12**
Date: 17-11-2024

Developer Name: Mayank Bansal

---

## Issue Ticket Number

#2217 
## Description
- Added unit test cases for authentication middleware inside `authenticate.test.js`
- Due to a missing import statement was getting an error while running tests. Fixed the error by adding the missing config statement in three files.

### Documentation Updated?

- [ ] Yes
- [x] No

### Under Feature Flag

- [ ] Yes
- [x] No

### Database Changes

- [ ] Yes
- [x] No

### Breaking Changes

- [ ] Yes
- [x] No

### Development Tested?

- [x] Yes
- [ ] No

## Test Coverage
<details>
<summary>Screenshot 1</summary>

added test cases

</details>

<details>
<summary>Screenshot 2</summary>

test coverage

</details>

### CODE REVIEW

**Comment by pankajjs** [comment]
Are tests not working without importing config because it is a global variable?
@MayankBansal12

**Comment by MayankBansal12** [comment]
@pankajjs I tried running it without the import, but I have been getting `config is not defined` error.

**Comment by pankajjs** [comment]
> @pankajjs I tried running it without the import, but I have been getting `config is not defined` error.

Then how is it running before? @MayankBansal12

File: test/unit/middlewares/authenticate.test.js [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+      process.env.NODE_ENV = originalEnv;
+    });
+
+    it("should handle unexpected errors gracefully", async function () {
+      req.cookies[config.get("userToken.cookieName")] = "validToken";
+      const verifyAuthTokenStub = Sinon.stub(authService, "verifyAuthToken").throws(new Error("Unexpected error"));
+
+      await authMiddleware(req, res, nextSpy);
+
+      expect(verifyAuthTokenStub.calledOnce).to.equal(true);
```
Comment (pankajjs) [inline test/unit/middlewares/authenticate.test.js:95]: Can you assert the thrown error?

File: test/unit/middlewares/authenticate.test.js [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+      expect(verifyAuthTokenStub.calledWith("validToken")).to.equal(true);
+      expect(retrieveUsersStub.calledOnce).to.equal(true);
+      expect(res.boom.forbidden.calledOnce).to.equal(true);
+      expect(res.boom.forbidden.firstCall.args[0]).to.equal("You are restricted from performing this action");
+      expect(nextSpy.notCalled).to.equal(true);
+    });
+
+    it("should deny access with invalid token", async function () {
+      req.cookies[config.get("userToken.cookieName")] = "invalidToken";
+      const verifyAuthTokenStub = Sinon.stub(authService, "verifyAuthToken").throws(new Error("Invalid token"));
```
Comment (pankajjs) [inline test/unit/middlewares/authenticate.test.js:66]: Same

File: test/unit/middlewares/authenticate.test.js [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+      expect(res.boom.unauthorized.notCalled).to.equal(true);
+      expect(res.boom.forbidden.notCalled).to.equal(true);
+    });
+
+    it("should deny restricted user access for non-GET requests", async function () {
+      req.cookies[config.get("userToken.cookieName")] = "validToken";
+      req.method = "POST";
+      const user = { id: "user123", roles: { restricted: true } };
+      const verifyAuthTokenStub = Sinon.stub(authService, "verifyAuthToken").returns({ userId: user.id });
+      const retrieveUsersStub = Sinon.stub(dataAccess, "retrieveUsers").resolves({ user });
```
Comment (pankajjs) [inline test/unit/middlewares/authenticate.test.js:52]: Assert the return value of these stubs

File: test/unit/middlewares/authenticate.test.js [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+  afterEach(function () {
+    Sinon.restore();
+  });
+
+  describe("Token Verification", function () {
+    it("should allow unrestricted user with valid token", async function () {
+      req.cookies[config.get("userToken.cookieName")] = "validToken";
+      const user = { id: "user123", roles: { restricted: false } };
+      const verifyAuthTokenStub = Sinon.stub(authService, "verifyAuthToken").returns({ userId: user.id });
+      const retrieveUsersStub = Sinon.stub(dataAccess, "retrieveUsers").resolves({ user });
```
Comment (pankajjs) [inline test/unit/middlewares/authenticate.test.js:35]: Same

**Comment by pankajjs** [comment]
Left few comments @MayankBansal12

**Comment by MayankBansal12** [comment]
> Left few comments @MayankBansal12

Will fix the issues mentioned

**Comment by MayankBansal12** [comment]
@pankajjs Added a new commit making the changes mentioned above. Have a look.

**Review by pankajjs — APPROVED** [review]
LGTM ✅

**Review by iamitprakash — APPROVED** [review]

**Pull request merged**